forwards_paper.tex

```diff
@@ -611,18 +629,327 @@ \subsection*{Tree sequence simplification}
+Roughly, the length of time for which a marginal tree has $k$ tips is
+$2N/k(k-1) = 2N(1/(k-1) - 1/k)$,
+and so the time to go from $N$ to $n$ tips is $2N(1/(n-1) - 1/(N-1))$.
+This implies that after running for time $T$ we expect to
+have around $n(T)$ roots, where $n(T) \approx 1 + 2N/(T+2)$.
+The total tree length over this time is
+$2 N \sum_{k=n(T)+1}^{N-1} 1/k$, which
+% is approximately
+% \begin{align*}
+%     2 N \log\left( \frac{N}{1 + 2 N/(T+2)} \right) .
+% \end{align*}
+leads to an upper bound on the number of edges of
+\begin{align}
+    \label{eqn:edge_bound}
+    2 N \left( 1 + 4 \log\left( \frac{NT}{T + 2 N} \right)\right) .
+\end{align}
+This holds up quite well in practice,
+as shown in Figure~\ref{fig:simplify_complexity}.
+% by comparison to
+% the number of edges actually required for 50 independent Wright--Fisher simulations,
+% as a function of time (obtained from Algorithm~\algref{W} with $s=1$)
+% is shown in Figure~\ref{fig:simplify_complexity}, compared to this prediction.
+
+\begin{figure}
+    \begin{center}
+        \includegraphics{sims/num_edges_50}
+        \includegraphics{sims/simplify_timings}
+    \end{center}
+    \caption{
+        \textbf{(A)}
+        Number of edges in the simplified tree sequence
+        for 50 Wright--Fisher simulations begun with no prior history,
+        as a function of number of generations.
+        Each line is one simulation, the black line gives the average,
+        and the dotted line is the upper bound of equation \eqref{eqn:edge_bound}.
+        \plr{Propose to add other figures as subfigures here; but better than this.}
+        \textbf{(B)}
+        Time required to simplify the first $k$ edges
+        of \plr{what tree sequence?},
+        plotted against $k$.
+        \textbf{(C)}
+        Time required to simplify
+        \plr{what tree sequence?}
+        down to a sample of size $n$,
+        plotted against $n$
+        (note the log scale;
+        the time scales logarithmically with $n$).
+        \label{fig:simplify_complexity}
+    }
+\end{figure}
+
+%% Above was edges; same holds for memory usage savings due to mutations
+If we were to add mutations in forwards time
+under the infinite-sites model with total mutation rate per generation $\mu$,
+there would also be around $\mu 2NT$ mutations (and the same number of sites).
+Since mutations fall as recombinations do on the marginal trees,
+adding them after simplification results in only around $\mu 4 N \log(2N)$ mutations,
+and the same savings as we obtained from edges above,
+even without considering the 
+computational burden of propagating mutations forwards across generations.
+
+
+%% simplify run time
+Above we have described memory requirements;
+how does the computation time required for simplification scale?
+Simply because it must process each edge,
+the simplification algorithm is linear in the number of edges of the input tree sequence.
```

Comment:
I'm a bit skeptical about this. Clearly it's linear in the number of edges, but we're assuming the cost of processing each edge is O(1) and this is ignoring the cost of computing overlaps. We do two things for each edge:

1. Find where it overlaps with existing ancestral material in A[c], and add this to an interval tree P. Can we assume that |A[c]| is O(1). Maybe we can for the forward simulation case, but I don't think this is true in general.
2. Find the union of all overlaps in P (I don't see how this can be O(1) either)

Anyway, I guess my point is I'm not totally convinced by this run time analysis. Perhaps we could think about the properties of what comes out of Alg W a bit and use this to get a better idea of the per-edge cost? Or we could fudge it and just say that the per-edge cost is typically small for these inputs.

Reply:
Hm, good point.  I was mostly trying to describe the figure, which says that *in practice* it seems to be linear.  We should at least change that to say "at least linear", and go on to say more.

I guess this depends on what we are scaling.  If we're talking about time to simplify the results from T generations of a given fwds simulation of a pop of size N, then it should be linear in T, because looking at more time does not affect the number of overlaps at a given node.   This is I think what we're really discussing here.  Keeping T fixed and increasing N may be different.

By what you say, it should scale with (number of edges) * (K log K), where K is the number of intervals per node, I think?